ambry-router/src/main/java/com/github/ambry/router/TtlUpdateManager.java

```diff
@@ -81,42 +83,64 @@ class TtlUpdateManager {
-        routerMetrics.ttlUpdateBlobNotOriginateLocalOperationRate.mark();
-      }
-      blobIds.add(blobId);
+    BlobId blobId = RouterUtils.getBlobIdFromString(blobIdStr, clusterMap);
+    if (blobId.getDatacenterId() != ClusterMap.UNKNOWN_DATACENTER_ID
+        && blobId.getDatacenterId() != clusterMap.getLocalDatacenterId()) {
+      routerMetrics.ttlUpdateBlobNotOriginateLocalOperationRate.mark();
+    }
+
+    // ChunkId list should not contain the metadata blobId itself.
+    while (chunkIdStrs != null && chunkIdStrs.contains(blobIdStr)) {
+      chunkIdStrs.remove(blobIdStr);
     }
-    if (blobIds.size() == 1) {
+
+    if (chunkIdStrs == null || chunkIdStrs.isEmpty()) {
+      // If there are no chunkIds, then its a simple blob, and we have to update only blob's ttl.
       TtlUpdateOperation ttlUpdateOperation =
-          new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobIds.get(0), serviceId, expiresAtMs,
+          new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobId, serviceId, expiresAtMs,
               time.milliseconds(), callback, time, futureResult, quotaChargeCallback);
       ttlUpdateOperations.add(ttlUpdateOperation);
-    } else {
-      BatchOperationCallbackTracker tracker =
-          new BatchOperationCallbackTracker(blobIds, futureResult, callback, quotaChargeCallback);
-      long operationTimeMs = time.milliseconds();
-      for (BlobId blobId : blobIds) {
-        TtlUpdateOperation ttlUpdateOperation =
-            new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, blobId, serviceId, expiresAtMs,
-                operationTimeMs, tracker.getCallback(blobId), time, BatchOperationCallbackTracker.DUMMY_FUTURE,
-                quotaChargeCallback);
-        ttlUpdateOperations.add(ttlUpdateOperation);
+      return;
+    }
+
+    // If we are here, that means the blob is a composite blob. So we will do a batch operation.
+    List<BlobId> chunkIds = new ArrayList<>();
+    for (String chunkIdStr : chunkIdStrs) {
+      BlobId chunkId = RouterUtils.getBlobIdFromString(chunkIdStr, clusterMap);
+      if (chunkId.getDatacenterId() != ClusterMap.UNKNOWN_DATACENTER_ID
+          && chunkId.getDatacenterId() != clusterMap.getLocalDatacenterId()) {
+        routerMetrics.ttlUpdateBlobNotOriginateLocalOperationRate.mark();
       }
+      chunkIds.add(chunkId);
+    }
+
+    BatchOperationCallbackTracker tracker =
+        new BatchOperationCallbackTracker(chunkIds, blobId, futureResult, callback, quotaChargeCallback,
+            (bId, callBack) -> {
+              TtlUpdateOperation ttlUpdateOperation =
+                  new TtlUpdateOperation(clusterMap, routerConfig, routerMetrics, bId, serviceId, expiresAtMs,
+                      time.milliseconds(), callBack, time, futureResult, quotaChargeCallback);
+              ttlUpdateOperations.add(ttlUpdateOperation);
+            });
+    long operationTimeMs = time.milliseconds();
+    for (BlobId chunkId : chunkIds) {
```

Comment:
You can combine this loop with loop at 120. There are too many iterations and this will not scale for terabyte blobs since there'd be 262K chunks or more

Reply:
done.